module/validation/approval_validator.go

```diff
@@ -0,0 +1,76 @@
+package validation
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+type approvalValidator struct {
+	state    protocol.State
+	verifier module.Verifier
+}
+
+func NewApprovalValidator(state protocol.State, verifier module.Verifier) *approvalValidator {
+	return &approvalValidator{
+		state:    state,
+		verifier: verifier,
+	}
+}
+
+func (v *approvalValidator) Validate(approval *flow.ResultApproval) error {
+	// check if we already have the block the approval pertains to
+	head, err := v.state.AtBlockID(approval.Body.BlockID).Head()
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return fmt.Errorf("failed to retrieve header for block %x: %w", approval.Body.BlockID, err)
+		}
+		return engine.NewOutdatedInputErrorf("no header for block: %v", approval.Body.BlockID)
```

Comment:
⚠️ 
I don't this always means that the input is outdated: If the replica is behind and hasn't heard about the block, we will also have `storage.ErrNotFound` error.  

I see two options: 
* (preferred) a dedicated sentinel error: `engine.UnverifiableInputError` 
* we could use `InvalidInputError` as a temporary replacement (I think we did the same for the receipt validator)

Reply:
We can't use `InvalidInputError` since it will fail, it still should add entry into mempool. In docs outdated error states that it can be used for value that cannot be validation by any reasons. But I can add `UnverifiableInputError`